src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp

```diff
@@ -4632,6 +4632,69 @@ class StubGenerator: public StubCodeGenerator {
     return entry;
   }
 
+  // a1 = r1 - string1 address
+  // a2 = r3 - string2 address
+  // result = r0 - return value. Already contains "false"
+  // cnt1 = r4 - amount of elements left to check
+  address generate_long_string_equals() {
+    Register a1 = r1, a2 = r3, result = r0, cnt1 = r4;
+    Label NOT_EQUAL, EQUAL, LOOP, SMALL_LOOP, POST_LOOP;
+    int loopThreshold = 4 * wordSize;
+
+    assert_different_registers(a1, a2, result, cnt1);
+
+    __ align(CodeEntryAlignment);
+
+    StubCodeMark mark(this, "StubRoutines", "long_string_equals");
+
+    address entry = __ pc();
+    __ enter();
+
+    // reset cnt1
+    __ add(cnt1, cnt1, loopThreshold);
+
+    // Main 32 byte comparison loop.
+    __ bind(LOOP);
+      __ ld1(v0, v1, __ T2D, Address(__ post(a1, loopThreshold)));
+      __ sub(cnt1, cnt1, loopThreshold);
+      __ ld1(v2, v3, __ T2D, Address(__ post(a2, loopThreshold)));
+      __ subs(zr, cnt1, loopThreshold);
+      __ eor(v0, __ T16B, v0, v2);
+      __ eor(v1, __ T16B, v1, v3);
+      __ orr(v0, __ T16B, v0, v1);
+      __ umov(rscratch1, v0, __ D, 0);
+      __ umov(rscratch2, v0, __ D, 1);
+      __ orr(rscratch1, rscratch1, rscratch2);
+      __ cbnz(rscratch1, NOT_EQUAL);
+      __ br(__ GE, LOOP);
+
```

Comment:
As I said before, we gain nothing by using Neon here.

Reply:
Much better:

```
+	__ ldp(r5, r6, Address(__ post(a1, wordSize * 2)));
+	__ ldp(rscratch1, rscratch2, Address(__ post(a2, wordSize * 2)));
+	__ cmp(r5, rscratch1);
+	__ ccmp(r6, rscratch2, 0, Assembler::EQ);
+	__ br(__ NE, NOT_EQUAL);
```

Reply:
We changed `ld1` into `ldp` and get the result as following, 

simple:
Benchmark |(size)| Mode| Cnt | Score| Error |Units
-------------------|------|-----|-----|-------|---------|-----
StringEquals.equal |45  |avgt  |5 | 6.105 | ? 0.635  |us/op
StringEquals.equal  |64  |avgt | 5  |7.226  |? 0.056  |us/op
StringEquals.equal  |91 | avgt  |5  |12.010  |? 0.375 | us/op
StringEquals.equal  |121  |avgt  |5  |14.772  |? 0.114 | us/op
StringEquals.equal  |181 | avgt  |5 | 21.468 | ? 0.676  |us/op
StringEquals.equal  |256 | avgt  |5  |28.942  |? 4.806  |us/op
StringEquals.equal | 512  |avgt | 5  |58.479  |? 5.918  |us/op
StringEquals.equal  |1024  |avgt  |5  |119.313 | ? 16.661 | us/op
 
ldp:
Benchmark |(size)| Mode| Cnt | Score| Error |Units
-------------------|------|-----|-----|-------|---------|-----
StringEquals.equal  |45  |avgt  |5  |6.449 | ? 0.202 |us/op
StringEquals.equal  |64  |avgt | 5  |7.367  |? 0.055 |us/op
StringEquals.equal  |91  |avgt  |5 | 9.984  |? 0.065 |us/op
StringEquals.equal | 121 | avgt | 5 | 12.540  |? 0.545| us/op
StringEquals.equal  |181  |avgt  |5 | 15.614  |? 0.280 |us/op
StringEquals.equal | 256  |avgt | 5  |19.346 | ? 0.243| us/op
StringEquals.equal | 512  |avgt  |5  |35.718 | ? 0.599 |us/op
StringEquals.equal  |1024  |avgt  |5  |67.846 | ? 0.439| us/op

neon:
Benchmark |(size)| Mode| Cnt | Score| Error |Units
-------------------|------|-----|-----|-------|---------|-----
StringEquals.equal  |45 | avgt | 5 | 5.883  |? 0.173 | us/op
StringEquals.equal | 64  |avgt  |5 | 6.737  |? 0.035  |us/op
StringEquals.equal | 91 | avgt  |5  |8.997  |? 0.215  |us/op
StringEquals.equal  |121 | avgt | 5 | 10.789  |? 0.386  |us/op
StringEquals.equal  |181  |avgt  |5  |14.063  |? 0.253  |us/op
StringEquals.equal  |256 | avgt  |5  |19.679 | ? 1.419  |us/op
StringEquals.equal  |512  |avgt  |5  |38.813  |? 1.378  |us/op
StringEquals.equal  |1024  |avgt  |5 | 77.769  |? 3.082 | us/op

From the results, we can see that, 

* for small size (45~181), the performance of `ldp` version is not as good as `neon/ ld1` version
* for big size, `ldp` version is better that `neon/ld1` version
* all versions (both `ldp` and `ld1`) are better that old `simple` version .
* I agree with you `ldp` version is better than `ld1` version at **last patch** because I used 
```
__ ldr(v0, __ Q, Address(__ post(a1, wordSize * 2))); 
__ ldr(v1, __ Q, Address(__ post(a2, wordSize * 2)));
```
at last patch. However, I use 
```
__ ld1(v0, v1, __ T2D, Address(__ post(a1, loopThreshold)));
__ ld1(v2, v3, __ T2D, Address(__ post(a2, loopThreshold)));

```
in recent patch. I think this change has fixed the problem here.